WordPress/gutenberg
# Patterns: Add renaming, duplication, and deletion options

Fixes: https://github.com/WordPress/gutenberg/issues/51919

Related:
- https://github.com/WordPress/gutenberg/issues/52107

**NOTE: This PR still needs some polishing and updates to ensure better a11y**

## What?

This PR allows for:
- renaming or deleting custom user patterns & custom template parts
- duplicating all patterns or template parts (custom, theme, or core)

### Known Issues / Follow Ups

- The stacking of snackbar notices is being worked on in a separate PR
- The flow when creating a pattern that returns via the "manage all" list view pages could be improved in a follow-up
- We likely need some keyboard shortcuts similar to the delete option for the renaming and duplication
    - If the patterns list is changed to a simple list the need for the keyboard event handling/shortcuts might be alleviated

> However, I'd reconsider the listbox / options implementation as it introduces unnecessary complex interaction for screen reader users. This i sjust a list of objects, visually presented in a grid. I'd recommend to use a standard list.

See: https://github.com/WordPress/gutenberg/issues/52253

## Why?

This improves management options for patterns and template parts and avoids convoluted flows to affect relatively minor changes such as renaming.

## How?

- Adds new options to each pattern's dropdown menu
- All patterns or template parts, have a "Duplicate" option now
- Custom user-created patterns or template parts have both "Rename" and "Delete" options

## Testing Instructions

1. Create some synced and unsynced patterns via the Site Editor > Patterns page
2. Test renaming your created patterns
3. Try out duplicating your patterns
4. Ensure you can delete your patterns
5. Navigate to one of the theme pattern categories e.g. Featured for TT3 theme
6. Confirm that you only have the "duplicate" option for a theme pattern and it works as expected
7. Navigate to one of the template part categories and find a core template part e.g. Header or Footer
8. Duplicate the core template part
9. Check that you can now also rename and delete this newly created custom template part

### Testing Instructions for Keyboard

TBA


## Screenshots or screencast 


https://github.com/WordPress/gutenberg/assets/60436221/41cdc2c8-f683-4db7-a19f-3298832fd4a1
